ironfish-cli/src/commands/ceremony.ts

```diff
@@ -0,0 +1,170 @@
+    const outputPath = path.join(tempDir, 'newParams')
+
+    let localHash: string | null = null
+
+    // Prompt for randomness
+    let randomness: string | null = await CliUx.ux.prompt(
+      "Provide some randomness to contribute to the ceremony. For more information on where this should come from and it's importance, please read [link_to_blog_or_randomness_info]. If none is provided, it will automatically be generated for you (press enter)",
+      { required: false },
+    )
+    randomness = randomness.length ? randomness : null
+
+    // Create the client and bind events
+    const client = new CeremonyClient({
+      host,
+      port,
+      logger: this.logger.withTag('ceremonyClient'),
+    })
+
+    client.onJoined.on(({ queueLocation }) => {
+      CliUx.ux.action.status = `Current position: ${queueLocation}`
+    })
+
+    client.onInitiateContribution.on(async ({ downloadLink, contributionNumber }) => {
+      CliUx.ux.action.stop()
+
+      this.log(`Starting contribution. You are contributor #${contributionNumber}`)
+
+      CliUx.ux.action.start(`Downloading the previous contribution to ${inputPath}`)
+
+      const fileHandle = await fsAsync.open(inputPath, 'w')
+
+      let response
+      try {
+        response = await axios.get(downloadLink, {
+          responseType: 'stream',
+          onDownloadProgress: (p: ProgressEvent) => {
+            this.log('loaded', p.loaded, 'total', p.total)
+          },
+        })
+      } catch (e) {
+        this.error(ErrorUtils.renderError(e))
+      }
+
+      await pipeline(response.data, fileHandle.createWriteStream())
+
+      CliUx.ux.action.stop()
+
+      CliUx.ux.action.start(`Contributing your randomness`)
+
+      localHash = await contribute(inputPath, outputPath, randomness)
+
+      CliUx.ux.action.stop()
+
+      CliUx.ux.action.start(`Waiting to upload your contribution`)
+
+      client.contributionComplete()
+    })
+
+    client.onInitiateUpload.on(async ({ uploadLink }) => {
+      CliUx.ux.action.stop()
+
+      CliUx.ux.action.start(`Uploading your contribution`)
+
+      const fileHandle = await fsAsync.open(outputPath, 'r')
+      const stat = await fsAsync.stat(outputPath)
+
+      try {
+        await axios.put(uploadLink, fileHandle.createReadStream(), {
+          maxBodyLength: 1000000000,
+          headers: {
+            'Content-Type': 'application/octet-stream',
+            'Content-Length': stat.size,
+          },
+        })
+      } catch (e) {
+        this.error(ErrorUtils.renderError(e))
+      }
+
+      CliUx.ux.action.stop()
+      client.uploadComplete()
+
+      CliUx.ux.action.start('Contribution uploaded. Waiting for server to verify')
+    })
+
+    client.onContributionVerified.on(({ hash, downloadLink, contributionNumber }) => {
+      CliUx.ux.action.stop()
+
+      if (hash !== localHash) {
+        this.error(
+          'Hashes do not match. Please contact the Iron Fish team with this error message.',
```

Comment:
maybe we should print out the hashes here so we have more info to go on if there's an error

Reply:
Changed this to print out the hashes.